### PR TITLE
slack-cli: init at 0.18.0

### PR DESCRIPTION
**Pull request by alyssais**
- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

slack-cli must be configured using the SLACK_CLI_TOKEN environment variable. Using `slack init` will not work because it tries to write to the Nix store.

### CODE REVIEW

**Comment by mpickering** [comment]
@GrahamcOfBorg build slack-cli

**Comment by mpickering** [comment]
Thanks, looks fine to me.

File: pkgs/tools/networking/slack-cli/default.nix
Comment (timokau) [inline pkgs/tools/networking/slack-cli/default.nix]: It is customary to add the version to the name: `name = "slack-cli-${version}"`.

File: pkgs/tools/networking/slack-cli/default.nix
Comment (timokau) [inline pkgs/tools/networking/slack-cli/default.nix]: This shouldn't be necessary. The standard build phase uses make, it does not need to be specified as a dependency.

File: pkgs/tools/networking/slack-cli/default.nix
Comment (timokau) [inline pkgs/tools/networking/slack-cli/default.nix]: Where do you actually use `make`?

File: pkgs/tools/networking/slack-cli/default.nix
Comment (timokau) [inline pkgs/tools/networking/slack-cli/default.nix]: Please quote shell variables, including `$out`.

**Review by timokau — CHANGES_REQUESTED** [review]

File: pkgs/tools/networking/slack-cli/default.nix
Comment (mpickering) [reply pkgs/tools/networking/slack-cli/default.nix]: The `makeWrapper` dependency is for the `installPhase` where `wrapProgram` is called.

Comment (timokau) [reply pkgs/tools/networking/slack-cli/default.nix]: Ooh that is a confusing name. I assumed it was a wrapper for make :grinning:

It should be a `nativeBuildInput` then. The difference is that `nativeBuildInputs` are used exclusively during packaging. The distinction is important for cross-compilation.

Alternatively you could leave it out of any `buildInput`s and just use `${makeWrapper}/bin/wrapProgram` in `installPhase`.

File: pkgs/tools/networking/slack-cli/default.nix
Comment (alyssais) [reply pkgs/tools/networking/slack-cli/default.nix]: It could be worked around by changing the location it saves to, but then a decision would have to be made where that should be. A wrapper script that warns if the environment variable is not set is a good idea though.

File: pkgs/tools/networking/slack-cli/default.nix
Comment (timokau) [reply pkgs/tools/networking/slack-cli/default.nix]: > fwiw

Yes that is because the distinction is relatively recent, many contributors aren't aware of it and it is not enforced in regular compilation (yet). But the right way to do it is to always put `makeWrapper` in `nativeBuildInputs`. Maybe some point in the future we will enforce the distinction for regular compilation.

For example when you want to compile for a raspberry pi on your regular x86 desktop, you will need a x86 version of `makeWrapper` but aarch versions of all the runtime dependencies.

> This wouldn't work.

Huh, I wonder how it works when it is in `buildInputs`. Anyways, that was a misunderstanding on my side. The way you do it currently is fine (except native vs. regular build inputs).

File: pkgs/tools/networking/slack-cli/default.nix
Comment (timokau) [reply pkgs/tools/networking/slack-cli/default.nix]: Sounds like that should be in the home directory. Where does it regularly save it?

But either way, your call. I don't use the application. Just silently failing without giving the user a hint for the reason is sub-optimal.

File: pkgs/tools/networking/slack-cli/default.nix
Comment (alyssais) [reply pkgs/tools/networking/slack-cli/default.nix]: Cool, thanks for the explanation :)

File: pkgs/tools/networking/slack-cli/default.nix
Comment (alyssais) [reply pkgs/tools/networking/slack-cli/default.nix]: It looks like it wants to save it to `etc`, which seems very odd to me, because then it can't be used by multiple users. I think that changing it only for nixpkgs would be confusing, so I'm going to leave it as-is, since I'm happy using the environment variable. That way if somebody else wants to use the file approach they can change it to be what they think makes sense.

Comment (timokau) [reply pkgs/tools/networking/slack-cli/default.nix]: Yes, that seems reasonable (unlike the upstream approach of saving user configuration to `/etc`). But then please add a wrapper that notifies the user and maybe add a comment explaining the alternative solution (in case somebody wants to improve it in the future).

Comment (alyssais) [reply pkgs/tools/networking/slack-cli/default.nix]: Will do.

File: pkgs/top-level/all-packages.nix
Comment (timokau) [inline pkgs/top-level/all-packages.nix]: Is there a reason to add all 3 of these to `all-packages.nix`?

File: pkgs/top-level/all-packages.nix
Comment (timokau) [inline pkgs/top-level/all-packages.nix]: Is there a reason to add all three of these to `all-packages.nix`?

File: pkgs/tools/networking/slack-cli/wrapper.nix
Comment (timokau) [inline pkgs/tools/networking/slack-cli/wrapper.nix]: Personally I'd just inline the wrapper into the main package. Any reason you chose this approach?

Comment (alyssais) [reply pkgs/tools/networking/slack-cli/wrapper.nix]: There is indeed! See https://github.com/NixOS/nixpkgs/pull/44102, but the tl;dr is that inlining the wrapper into the main package breaks `overrideAttrs`, so as a user of the package you can't eg add patches.

File: pkgs/top-level/all-packages.nix
Comment (Infinisil) [reply pkgs/top-level/all-packages.nix]: A reason one might prefer a wrapper derivation is when the unwrapped one takes a long time to rebuild but the wrapper should have a quick turnaround time. Changing something in the wrapper doesn't affect the unwrapped one at all and can therefore be very quick.

Comment (alyssais) [reply pkgs/top-level/all-packages.nix]: https://github.com/NixOS/nixpkgs/pull/45055#discussion_r210268395

Comment (alyssais) [reply pkgs/top-level/all-packages.nix]: @Infinisil what do you think of https://github.com/NixOS/nixpkgs/pull/45055#discussion_r210268395?

File: pkgs/tools/networking/slack-cli/wrapper.nix
Comment (Infinisil) [reply pkgs/tools/networking/slack-cli/wrapper.nix]: It doesn't need to be 2 packages at all: Just "insert" the wrapper into the original derivation like this: `ln -s ${wrapper}/bin/slack-cli $out/bin/slack-cli`, while moving the original slack-cli binary to `$out/bin/.slack-cli-wrapped_`

Comment (alyssais) [reply pkgs/tools/networking/slack-cli/wrapper.nix]: Oh, I see what you're saying! I'll do that.

File: pkgs/tools/networking/slack-cli/default.nix
Comment (Infinisil) [inline pkgs/tools/networking/slack-cli/default.nix]: You can use

```nix
let
  wrapper = writeShellScriptBin "slack" ''
    ...
    exec "${slack-cli}/bin/.slack-wrapped"
  '';

  slack-cli = stdenv.mkDerivation rec {
    ...
  };

in slack-cli
```

(or similar) to not have to use this dirname hack

Comment (alyssais) [reply pkgs/tools/networking/slack-cli/default.nix]: I tried this! (Or something very similar.) It was infinitely recursive though. Was I doing something wrong? It makes sense to me I think – `wrapper` takes `slack-cli`'s output path as an input, and `slack-cli`'s output path depends on all of its inputs, one of which is `wrapper`.

Comment (alyssais) [reply pkgs/tools/networking/slack-cli/default.nix]: @Infinisil that does work (just pushed), although it puts me in shell-escape hell…

Comment (Infinisil) [reply pkgs/tools/networking/slack-cli/default.nix]: Yeah that is a bit of a disadvantage..

Comment (alyssais) [reply pkgs/tools/networking/slack-cli/default.nix]: What do you think then? I think that, having seen it, I actually prefer this version. Even if it was a bit annoying to write, it’s less hacky.

Comment (Infinisil) [reply pkgs/tools/networking/slack-cli/default.nix]: Agreed!

**Comment by timokau** [comment]
LGTM, merge pending ofBorg's approval.
@GrahamcOfBorg build slack-cli

Thank you for packaging this and responding to the many nitpicks. Always happy to see a new entry in `maintainers.nix` :)

**Review by timokau — APPROVED** [review]

**Comment by timokau** [comment]
By the way we'll need to squash this for merge. I can just use githubs squash-and-merge feature, but in case you prefer to squash it yourself now's the time.

**Comment by timokau** [comment]
Since ofBorg approves, I don't think githubs squash-and-merge has any disadvantages and I want to merge this before going to sleep and potentially forgetting about this tomorrow, I'll just merge now :)

**Comment by alyssais** [comment]
Thanks for all the reviewing! I learned a lot :)

**Comment by Infinisil** [comment]
@timokau I prefer not to use squash-and-merge when the change should be in multiple commits, which would've been the case here: The maintainers update and the package init.

**Comment by timokau** [comment]
@Infinisil yes, but in this case it was a lot of "fixup/small change" commits and I think the maintainers upgrade was not actually in a separate commit. For what its worth, I pressed the wrong button anyways and did a regular merge :sweat_smile: 

But all the commits here can reasonable stand by themselves, so I think it is fine either way. I would just have preferred not to have multiple very small commits that don't add much value clutter the history.

**Pull request merged**
